File: client-sdks/sockudo-swift/Sources/SockudoSwift/Channels.swift
import Foundation
import Sodium

public class Channel: @unchecked Sendable {
  public let name: String
  unowned let client: SockudoClient
  let dispatcher: EventDispatcher
  var isSubscribed = false
  var subscriptionPending = false
  var subscriptionCancelled = false
  var subscriptionCount: Int?
  var tagsFilter: FilterNode?
  var deltaSettings: ChannelDeltaSettings?
  var eventsFilter: [String]?
  var rewind: SubscriptionRewind?
  var annotationSubscribe = false

  init(name: String, client: SockudoClient) {
    self.name = name
    self.client = client
    self.dispatcher = EventDispatcher { event, _ in
      Logger.debug("No callbacks on \(name) for \(event)")
    }
  }

  @discardableResult
  public func on(_ eventName: String, callback: @escaping (Any?, EventMetadata?) -> Void)
    -> EventBindingToken
  {
    dispatcher.bind(eventName, callback: callback)
  }

  @discardableResult
  public func bind(_ eventName: String, callback: @escaping (Any?, EventMetadata?) -> Void)
    -> EventBindingToken
  {
    on(eventName, callback: callback)
  }

  @discardableResult
  public func onGlobal(_ callback: @escaping (String, Any?) -> Void) -> EventBindingToken {
    dispatcher.bindGlobal(callback)
  }

  @discardableResult
  public func bindGlobal(_ callback: @escaping (String, Any?) -> Void) -> EventBindingToken {
    onGlobal(callback)
  }

  public func off(eventName: String? = nil, token: EventBindingToken? = nil) {
    dispatcher.unbind(eventName: eventName, token: token)
  }

  public func unbind(eventName: String? = nil, token: EventBindingToken? = nil) {
    off(eventName: eventName, token: token)
  }

  public func unbindAll() {
    dispatcher.unbind()
  }

  public func setDeltaSettings(_ settings: ChannelDeltaSettings?) {
    deltaSettings = settings
  }

  public func trigger(event: String, data: Any) throws -> Bool {
    guard event.hasPrefix("client-") else {
      throw SockudoError.badEventName("Event '\(event)' does not start with 'client-'")
    }
    if isSubscribed == false {
      Logger.warn("Client event triggered before channel subscription succeeded")
    }
    return try client.sendEvent(name: event, data: data, channel: name)
  }

  func authorize(
    socketID: String,
    completion: @escaping @Sendable (Result<ChannelAuthorizationData, Error>) -> Void
  ) {
    completion(
      .success(ChannelAuthorizationData(auth: "", channelData: nil, sharedSecret: nil)))
  }

  func subscribeIfPossible() {
    if subscriptionPending, subscriptionCancelled {
      subscriptionCancelled = false
    } else if subscriptionPending == false, client.connectionState == .connected {
      subscribe()
    }
  }

  func subscribe() {
    guard isSubscribed == false else { return }
    subscriptionPending = true
    subscriptionCancelled = false
    authorize(socketID: client.socketID ?? "") { [weak self] result in
      guard let self else { return }
      switch result {
      case .failure(let error):
        self.subscriptionPending = false
        self.dispatcher.emit(
          self.client.p.event("subscription_error"),
          data: [
            "type": "AuthError",
            "error": error.localizedDescription,
          ])
      case .success(let data):
        var payload: [String: Any] = [
          "auth": data.auth,
          "channel": self.name,
        ]
        if let channelData = data.channelData {
          payload["channel_data"] = channelData
        }
        if let filter = self.tagsFilter, let filterJSON = try? JSON.encodeData(filter),
          let json = try? JSON.decode(filterJSON)
        {
          payload["tags_filter"] = json
        }
        if let deltaSettings = self.deltaSettings {
          payload["delta"] = deltaSettings.subscriptionValue()
        }
        if let eventsFilter = self.eventsFilter {
          payload["events"] = eventsFilter
        }
        if let rewind = self.rewind {
          payload["rewind"] = rewind.subscriptionValue()
        }
        if self.annotationSubscribe {
          payload["modes"] = ["SUBSCRIBE", "ANNOTATION_SUBSCRIBE"]
        }
        do {
          _ = try self.client.sendEvent(
            name: self.client.p.event("subscribe"), data: payload, channel: nil)
        } catch {
          self.subscriptionPending = false
          self.dispatcher.emit(
            self.client.p.event("subscription_error"),
            data: [
              "type": "ConnectionError",
              "error": error.localizedDescription,
            ])
        }
      }
    }
  }

  func unsubscribe() {
    isSubscribed = false
    _ = try? client.sendEvent(
      name: client.p.event("unsubscribe"), data: ["channel": name], channel: nil)
  }

  func disconnect() {
    isSubscribed = false
    subscriptionPending = false
  }

  func handle(event: SockudoEvent) {
    let p = client.p
    if event.event == p.internal("subscription_succeeded") {
      subscriptionPending = false
      isSubscribed = true
      if subscriptionCancelled {
        client.unsubscribe(name)
      } else {
        dispatcher.emit(p.event("subscription_succeeded"), data: event.data)
      }
    } else if event.event == p.internal("subscription_count") {
      if let data = event.data as? [String: Any],
        let count = data["subscription_count"] as? Int
      {
        subscriptionCount = count
      }
      dispatcher.emit(p.event("subscription_count"), data: event.data)
    } else if event.event == p.internal("message"),
      let data = event.data as? [String: Any],
      data["action"] as? String == "message.summary"
    {
      dispatcher.emit("message.summary", data: data, metadata: EventMetadata(userID: event.userID))
    } else if event.event == p.internal("annotation"),
      let data = event.data as? [String: Any],
      let action = data["action"] as? String
    {
      dispatcher.emit(action, data: data, metadata: EventMetadata(userID: event.userID))
    } else if p.isInternalEvent(event.event) == false {
      dispatcher.emit(
        event.event, data: event.data, metadata: EventMetadata(userID: event.userID))
    }
  }

  public func publishAnnotation(
    messageSerial: String,
    annotation: PublishAnnotationRequest,
    completion: @escaping @Sendable (Result<PublishAnnotationResponse, Error>) -> Void
  ) {
    client.publishAnnotation(
      channelName: name,
      messageSerial: messageSerial,
      annotation: annotation,
      completion: completion
    )
  }

  public func deleteAnnotation(
    messageSerial: String,
    annotationSerial: String,
    socketID: String? = nil,
    completion: @escaping @Sendable (Result<DeleteAnnotationResponse, Error>) -> Void
  ) {
    client.deleteAnnotation(
      channelName: name,
      messageSerial: messageSerial,
      annotationSerial: annotationSerial,
      socketID: socketID,
      completion: completion
    )
  }

  public func listAnnotations(
    messageSerial: String,
    params: AnnotationEventsParams = .init(),
    completion: @escaping @Sendable (Result<AnnotationEventsPage, Error>) -> Void
  ) {
    client.listAnnotations(
      channelName: name,
      messageSerial: messageSerial,
      params: params,
      completion: completion
    )
  }
}

public class PrivateChannel: Channel, @unchecked Sendable {
  override func authorize(
    socketID: String,
    completion: @escaping @Sendable (Result<ChannelAuthorizationData, Error>) -> Void
  ) {
    client.config.channelAuthorizer(
      ChannelAuthorizationRequest(socketID: socketID, channelName: name),
      completion
    )
  }
}

public final class PresenceMembers: @unchecked Sendable {
  private(set) var members: [String: AnyHashable] = [:]
  public private(set) var count = 0
  public private(set) var myID: String?
  public private(set) var me: PresenceMember?

  public func member(id: String) -> PresenceMember? {
    guard let info = members[id] else { return nil }
    return PresenceMember(id: id, info: info)
  }

  public func forEach(_ body: (PresenceMember) -> Void) {
    for (id, info) in members {
      body(PresenceMember(id: id, info: info))
    }
  }

  func setMyID(_ id: String) {
    myID = id
  }

  func applySubscriptionData(_ data: [String: Any]) {
    let presence = data["presence"] as? [String: Any]
    let hash = presence?["hash"] as? [String: AnyHashable] ?? [:]
    members = hash
    count = presence?["count"] as? Int ?? hash.count
    if let myID {
      me = member(id: myID)
    }
  }

  func add(_ data: [String: Any]) -> PresenceMember? {
    guard let userID = data["user_id"] as? String else { return nil }
    let info = data["user_info"] as? AnyHashable
    if members[userID] == nil {
      count += 1
    }
    members[userID] = info
    return PresenceMember(id: userID, info: info)
  }

  func remove(_ data: [String: Any]) -> PresenceMember? {
    guard let userID = data["user_id"] as? String,
      let info = members.removeValue(forKey: userID)
    else {
      return nil
    }
    count = max(0, count - 1)
    return PresenceMember(id: userID, info: info)
  }

  func reset() {
    members.removeAll()
    count = 0
    myID = nil
    me = nil
  }
}

public final class PresenceChannel: PrivateChannel, @unchecked Sendable {
  public let members = PresenceMembers()

  override func authorize(
    socketID: String,
    completion: @escaping @Sendable (Result<ChannelAuthorizationData, Error>) -> Void
  ) {
    super.authorize(socketID: socketID) { [weak self] result in
      guard let self else { return }
      switch result {
      case .failure:
        completion(result)
      case .success(let data):
        if let channelData = data.channelData,
          let channelDataAny = try? JSON.decodeString(channelData),
          let dictionary = channelDataAny as? [String: Any],
          let userID = dictionary["user_id"] as? String
        {
          self.members.setMyID(userID)
          completion(.success(data))
          return
        }

        if let userID = self.client.user.userID {
          self.members.setMyID(userID)
          completion(.success(data))
        } else {
          completion(
            .failure(
              SockudoError.authFailure(
                statusCode: nil,
                message: "Invalid auth response for presence channel '\(self.name)'"
              )))
        }
      }
    }
  }

  override func handle(event: SockudoEvent) {
    let p = client.p
    if event.event == p.internal("subscription_succeeded") {
      subscriptionPending = false
      isSubscribed = true
      if subscriptionCancelled {
        client.unsubscribe(name)
      } else if let data = event.data as? [String: Any] {
        members.applySubscriptionData(data)
        dispatcher.emit(p.event("subscription_succeeded"), data: members)
      }
    } else if event.event == p.internal("subscription_count") {
      super.handle(event: event)
    } else if event.event == p.internal("member_added") {
      if let data = event.data as? [String: Any], let member = members.add(data) {
        dispatcher.emit(p.event("member_added"), data: member)
      }
    } else if event.event == p.internal("member_removed") {
      if let data = event.data as? [String: Any], let member = members.remove(data) {
        dispatcher.emit(p.event("member_removed"), data: member)
      }
    } else if event.event == p.internal("message"),
      let data = event.data as? [String: Any],
      data["action"] as? String == "message.summary"
    {
      dispatcher.emit("message.summary", data: data, metadata: EventMetadata(userID: event.userID))
    } else if event.event == p.internal("annotation"),
      let data = event.data as? [String: Any],
      let action = data["action"] as? String
    {
      dispatcher.emit(action, data: data, metadata: EventMetadata(userID: event.userID))
    } else if p.isInternalEvent(event.event) == false {
      dispatcher.emit(
        event.event, data: event.data, metadata: EventMetadata(userID: event.userID))
    }
  }

  override func disconnect() {
    members.reset()
    super.disconnect()
  }

  public func history(
    _ params: PresenceHistoryParams = .init(),
    completion: @escaping @Sendable (Result<PresenceHistoryPage, Error>) -> Void
  ) {
    client.fetchPresenceHistory(channelName: name, params: params, completion: completion)
  }

  public func channelHistory(
    _ params: ChannelHistoryParams = .init(),
    completion: @escaping @Sendable (Result<ChannelHistoryPage, Error>) -> Void
  ) {
    client.fetchChannelHistory(channelName: name, params: params, completion: completion)
  }

  public func getMessage(
    _ messageSerial: String,
    completion: @escaping @Sendable (Result<[String: Any], Error>) -> Void
  ) {
    client.fetchLatestMessage(channelName: name, messageSerial: messageSerial, completion: completion)
  }

  public func getMessageVersions(
    _ messageSerial: String,
    params: MessageVersionsParams = .init(),
    completion: @escaping @Sendable (Result<MessageVersionsPage, Error>) -> Void
  ) {
    client.fetchMessageVersions(
      channelName: name,
      messageSerial: messageSerial,
      params: params,
      completion: completion
    )
  }

  public func snapshot(
    _ params: PresenceSnapshotParams = .init(),
    completion: @escaping @Sendable (Result<PresenceSnapshot, Error>) -> Void
  ) {
    client.fetchPresenceSnapshot(channelName: name, params: params, completion: completion)
  }
}

public final class EncryptedChannel: PrivateChannel, @unchecked Sendable {
  private var sharedSecret: Bytes?
  private let sodium = Sodium()

  override func authorize(
    socketID: String,
    completion: @escaping @Sendable (Result<ChannelAuthorizationData, Error>) -> Void
  ) {
    super.authorize(socketID: socketID) { [weak self] result in
      guard let self else { return }
      switch result {
      case .failure:
        completion(result)
      case .success(let data):
        guard let secret = data.sharedSecret, let decoded = Data(base64Encoded: secret)
        else {
          completion(
            .failure(
              SockudoError.authFailure(
                statusCode: nil,
                message:
                  "No shared_secret key in auth payload for encrypted channel: \(self.name)"
              )))
          return
        }
        self.sharedSecret = Array(decoded)
        completion(
          .success(
            ChannelAuthorizationData(
              auth: data.auth, channelData: data.channelData, sharedSecret: nil)))
      }
    }
  }

  override public func trigger(event: String, data: Any) throws -> Bool {
    throw SockudoError.unsupportedFeature(
      "Client events are not currently supported for encrypted channels")
  }

  override func handle(event: SockudoEvent) {
    let p = client.p
    if p.isInternalEvent(event.event) || p.isPlatformEvent(event.event) {
      super.handle(event: event)
      return
    }
    guard
      let secret = sharedSecret,
      let payload = event.data as? [String: Any],
      let ciphertext = payload["ciphertext"] as? String,
      let nonce = payload["nonce"] as? String,
      let ciphertextData = Data(base64Encoded: ciphertext),
      let nonceData = Data(base64Encoded: nonce)
    else {
      Logger.error("Unexpected format for encrypted event on \(name)")
      return
    }

    guard
      let decrypted = sodium.secretBox.open(
        authenticatedCipherText: Array(ciphertextData), secretKey: secret,
        nonce: Array(nonceData))
    else {
      super.authorize(socketID: client.socketID ?? "") { [weak self] result in
        guard let self else { return }
        if case .success(let authData) = result,
          let sharedSecret = authData.sharedSecret,
          let refreshedData = Data(base64Encoded: sharedSecret)
        {
          self.sharedSecret = Array(refreshedData)
          self.handle(event: event)
        }
      }
      return
    }

    let bytes = Data(decrypted)
    let value = (try? JSON.decode(bytes)) ?? String(decoding: bytes, as: UTF8.self)
    dispatcher.emit(event.event, data: value)
  }
}

File: client-sdks/sockudo-kotlin/lib/src/main/kotlin/io/sockudo/client/Channels.kt
package io.sockudo.client

import com.iwebpp.crypto.TweetNaclFast
import java.util.Base64

open class SockudoChannel internal constructor(
    val name: String,
    internal val client: SockudoClient,
) {
    internal val dispatcher = EventDispatcher { event, _ ->
        SockudoLogger.debug("No callbacks on $name for $event")
    }

    var isSubscribed: Boolean = false
        internal set
    var subscriptionPending: Boolean = false
        internal set
    var subscriptionCancelled: Boolean = false
        internal set
    var subscriptionCount: Int? = null
        internal set
    var filter: FilterNode? = null
    var deltaSettings: ChannelDeltaSettings? = null
    var eventsFilter: List<String>? = null
    var rewind: SubscriptionRewind? = null
    var annotationSubscribe: Boolean = false

    fun on(eventName: String, callback: (Any?, EventMetadata?) -> Unit): EventBindingToken =
        dispatcher.bind(eventName, callback)

    fun bind(eventName: String, callback: (Any?, EventMetadata?) -> Unit): EventBindingToken = on(eventName, callback)

    fun onGlobal(callback: (String, Any?) -> Unit): EventBindingToken = dispatcher.bindGlobal(callback)

    fun bindGlobal(callback: (String, Any?) -> Unit): EventBindingToken = onGlobal(callback)

    fun off(eventName: String? = null, token: EventBindingToken? = null) {
        dispatcher.unbind(eventName, token)
    }

    fun unbind(eventName: String? = null, token: EventBindingToken? = null) = off(eventName, token)

    fun unbindAll() {
        dispatcher.unbind()
    }

    open fun trigger(event: String, data: Any): Boolean {
        require(event.startsWith("client-")) {
            throw SockudoException.BadEventName("Event '$event' does not start with 'client-'")
        }
        if (!isSubscribed) {
            SockudoLogger.warn("Client event triggered before channel subscription succeeded")
        }
        return client.sendEvent(event, data, name)
    }

    internal open suspend fun authorize(socketId: String): ChannelAuthorizationData =
        ChannelAuthorizationData(auth = "")

    internal fun subscribeIfPossible() {
        if (subscriptionPending && subscriptionCancelled) {
            subscriptionCancelled = false
        } else if (!subscriptionPending && client.connectionState == ConnectionState.CONNECTED) {
            client.launchSubscription { subscribe() }
        }
    }

    internal suspend fun subscribe() {
        if (isSubscribed) {
            return
        }

        subscriptionPending = true
        subscriptionCancelled = false

        try {
            val auth = authorize(client.socketId.orEmpty())
            val payload = linkedMapOf<String, Any?>(
                "auth" to auth.auth,
                "channel" to name,
            )
            auth.channelData?.let { payload["channel_data"] = it }
            filter?.let { payload["tags_filter"] = JsonSupport.fromJsonElement(JsonSupport.toJsonElement(it)) }
            deltaSettings?.let { payload["delta"] = it.subscriptionValue() }
            eventsFilter?.let { payload["events"] = it }
            rewind?.let { payload["rewind"] = it.subscriptionValue() }
            if (annotationSubscribe) {
                payload["modes"] = listOf("SUBSCRIBE", "ANNOTATION_SUBSCRIBE")
            }
            client.sendEvent(client.p.event("subscribe"), payload, null)
        } catch (error: Throwable) {
            subscriptionPending = false
            dispatcher.emit(
                client.p.event("subscription_error"),
                mapOf(
                    "type" to "AuthError",
                    "error" to (error.message ?: "Unknown auth error"),
                ),
            )
        }
    }

    suspend fun channelHistory(params: ChannelHistoryParams = ChannelHistoryParams()): ChannelHistoryPage =
        client.fetchChannelHistory(name, params)

    suspend fun getMessage(messageSerial: String): Map<String, Any?> =
        client.fetchLatestMessage(name, messageSerial)

    suspend fun getMessageVersions(
        messageSerial: String,
        params: MessageVersionsParams = MessageVersionsParams(),
    ): MessageVersionsPage = client.fetchMessageVersions(name, messageSerial, params)

    suspend fun publishAnnotation(
        messageSerial: String,
        annotation: PublishAnnotationRequest,
    ): PublishAnnotationResponse = client.publishAnnotation(name, messageSerial, annotation)

    suspend fun deleteAnnotation(
        messageSerial: String,
        annotationSerial: String,
        socketId: String? = null,
    ): DeleteAnnotationResponse = client.deleteAnnotation(name, messageSerial, annotationSerial, socketId)

    suspend fun listAnnotations(
        messageSerial: String,
        params: AnnotationEventsParams = AnnotationEventsParams(),
    ): AnnotationEventsPage = client.listAnnotations(name, messageSerial, params)

    internal open fun unsubscribe() {
        isSubscribed = false
        client.sendEvent(client.p.event("unsubscribe"), mapOf("channel" to name), null)
    }

    internal open fun disconnect() {
        isSubscribed = false
        subscriptionPending = false
    }

    internal open fun handle(event: SockudoEvent) {
        val p = client.p
        when {
            event.event == p.internal_("subscription_succeeded") -> {
                subscriptionPending = false
                isSubscribed = true
                if (subscriptionCancelled) {
                    client.unsubscribe(name)
                } else {
                    dispatcher.emit(p.event("subscription_succeeded"), event.data)
                }
            }

            event.event == p.internal_("subscription_count") -> {
                subscriptionCount = (event.data as? Map<*, *>)?.get("subscription_count") as? Int
                dispatcher.emit(p.event("subscription_count"), event.data)
            }

            event.event == p.internal_("message") -> {
                val data = event.data as? Map<*, *>
                if (data?.get("action") == "message.summary") {
                    dispatcher.emit("message.summary", event.data)
                }
            }

            event.event == p.internal_("annotation") -> {
                val data = event.data as? Map<*, *>
                val action = data?.get("action") as? String
                if (action != null) {
                    dispatcher.emit(action, event.data)
                }
            }

            else -> {
                if (!p.isInternalEvent(event.event)) {
                    dispatcher.emit(event.event, event.data, EventMetadata(userId = event.userId))
                }
            }
        }
    }
}

open class PrivateChannel internal constructor(
    name: String,
    client: SockudoClient,
) : SockudoChannel(name, client) {
    override suspend fun authorize(socketId: String): ChannelAuthorizationData =
        client.config.channelAuthorizer.authorize(ChannelAuthorizationRequest(socketId, name))
}

class PresenceMembers {
    private val members = linkedMapOf<String, Any?>()

    var count: Int = 0
        internal set
    var myId: String? = null
        internal set
    var me: PresenceMember? = null
        internal set

    fun member(id: String): PresenceMember? = members[id]?.let { PresenceMember(id, it) }

    fun forEach(body: (PresenceMember) -> Unit) {
        members.forEach { (id, info) -> body(PresenceMember(id, info)) }
    }

    internal fun rememberMyId(id: String) {
        myId = id
    }

    internal fun applySubscriptionData(data: Map<String, Any?>) {
        val presence = data["presence"] as? Map<*, *>
        val hash =
            (presence?.get("hash") as? Map<*, *>)?.entries?.associate { it.key.toString() to it.value } ?: emptyMap()
        members.clear()
        members.putAll(hash)
        count = (presence?.get("count") as? Number)?.toInt() ?: hash.size
        me = myId?.let(::member)
    }

    internal fun add(data: Map<String, Any?>): PresenceMember? {
        val userId = data["user_id"] as? String ?: return null
        val info = data["user_info"]
        if (!members.containsKey(userId)) {
            count += 1
        }
        members[userId] = info
        return PresenceMember(userId, info)
    }

    internal fun remove(data: Map<String, Any?>): PresenceMember? {
        val userId = data["user_id"] as? String ?: return null
        val info = members.remove(userId) ?: return null
        count = maxOf(0, count - 1)
        return PresenceMember(userId, info)
    }

    internal fun reset() {
        members.clear()
        count = 0
        myId = null
        me = null
    }
}

class PresenceChannel internal constructor(
    name: String,
    client: SockudoClient,
) : PrivateChannel(name, client) {
    val members = PresenceMembers()

    override suspend fun authorize(socketId: String): ChannelAuthorizationData {
        val response = super.authorize(socketId)
        val channelData = response.channelData
        if (channelData != null) {
            val parsed = JsonSupport.fromJsonElement(JsonSupport.decode(channelData)) as? Map<*, *>
            val userId = parsed?.get("user_id") as? String
            if (userId != null) {
                members.rememberMyId(userId)
                return response
            }
        }

        client.user.userId?.let {
            members.rememberMyId(it)
            return response
        }

        throw SockudoException.AuthFailure(
            null,
            "Invalid auth response for presence channel '$name'",
        )
    }

    override fun handle(event: SockudoEvent) {
        val p = client.p
        when {
            event.event == p.internal_("subscription_succeeded") -> {
                subscriptionPending = false
                isSubscribed = true
                if (subscriptionCancelled) {
                    client.unsubscribe(name)
                } else {
                    val payload = event.data as? Map<String, Any?> ?: emptyMap()
                    members.applySubscriptionData(payload)
                    dispatcher.emit(p.event("subscription_succeeded"), members)
                }
            }

            event.event == p.internal_("subscription_count") -> super.handle(event)
            event.event == p.internal_("member_added") -> {
                val payload = event.data as? Map<String, Any?> ?: return
                members.add(payload)?.let { dispatcher.emit(p.event("member_added"), it) }
            }

            event.event == p.internal_("member_removed") -> {
                val payload = event.data as? Map<String, Any?> ?: return
                members.remove(payload)?.let { dispatcher.emit(p.event("member_removed"), it) }
            }

            else -> {
                if (!p.isInternalEvent(event.event)) {
                    dispatcher.emit(event.event, event.data, EventMetadata(userId = event.userId))
                }
            }
        }
    }

    override fun disconnect() {
        members.reset()
        super.disconnect()
    }

    suspend fun history(params: PresenceHistoryParams = PresenceHistoryParams()): PresenceHistoryPage =
        client.fetchPresenceHistory(name, params)

    suspend fun snapshot(params: PresenceSnapshotParams = PresenceSnapshotParams()): PresenceSnapshot =
        client.fetchPresenceSnapshot(name, params)
}

class EncryptedChannel internal constructor(
    name: String,
    client: SockudoClient,
) : PrivateChannel(name, client) {
    private var sharedSecret: ByteArray? = null

    override suspend fun authorize(socketId: String): ChannelAuthorizationData {
        val response = super.authorize(socketId)
        val secret = response.sharedSecret
            ?: throw SockudoException.AuthFailure(
                null,
                "No shared_secret key in auth payload for encrypted channel: $name",
            )
        sharedSecret = java.util.Base64.getDecoder().decode(secret)
        return response.copy(sharedSecret = null)
    }

    override fun trigger(event: String, data: Any): Boolean {
        throw SockudoException.UnsupportedFeature(
            "Client events are not currently supported for encrypted channels",
        )
    }

    internal override fun handle(event: SockudoEvent) {
        if (client.p.isInternalEvent(event.event) || client.p.isPlatformEvent(event.event)) {
            super.handle(event)
            return
        }

        val secret = sharedSecret
        if (secret == null) {
            SockudoLogger.debug("Received encrypted event before key has been retrieved from the auth endpoint")
            return
        }

        val payload = event.data as? Map<*, *> ?: run {
            SockudoLogger.error("Unexpected format for encrypted event on $name")
            return
        }
        val cipherText = (payload["ciphertext"] as? String)?.let(Base64.getDecoder()::decode)
        val nonce = (payload["nonce"] as? String)?.let(Base64.getDecoder()::decode)
        if (cipherText == null || nonce == null) {
            SockudoLogger.error("Unexpected format for encrypted event on $name")
            return
        }
        if (cipherText.size < TweetNaclFast.SecretBox.overheadLength) {
            SockudoLogger.error("Expected encrypted event ciphertext length to be ${TweetNaclFast.SecretBox.overheadLength}, got: ${cipherText.size}")
            return
        }
        if (nonce.size < TweetNaclFast.SecretBox.nonceLength) {
            SockudoLogger.error("Expected encrypted event nonce length to be ${TweetNaclFast.SecretBox.nonceLength}, got: ${nonce.size}")
            return
        }

        val box = TweetNaclFast.SecretBox(secret)
        var bytes = box.open(cipherText, nonce)
        if (bytes == null) {
            SockudoLogger.debug("Failed to decrypt an event. Fetching a new key from the auth endpoint...")
            client.launchSubscription {
                runCatching { authorize(client.socketId.orEmpty()) }
                    .onSuccess {
                        val refreshedSecret = sharedSecret ?: return@onSuccess
                        bytes = TweetNaclFast.SecretBox(refreshedSecret).open(cipherText, nonce)
                        if (bytes == null) {
                            SockudoLogger.error("Failed to decrypt event with new key. Dropping encrypted event")
                            return@onSuccess
                        }
                        emitDecrypted(event.event, bytes!!)
                    }.onFailure {
                        SockudoLogger.error("Failed to fetch new encryption key: ${it.message}")
                    }
            }
            return
        }

        emitDecrypted(event.event, bytes)
    }

    private fun emitDecrypted(eventName: String, bytes: ByteArray) {
        val raw = bytes.decodeToString()
        val payload = runCatching { JsonSupport.fromJsonElement(JsonSupport.decode(raw)) }.getOrElse { raw }
        dispatcher.emit(eventName, payload)
    }
}
